feat: remember stopped extensions across restart

deboer-tim opened the pull request:
### What does this PR do?

Introduces a new hidden configuration value 'extensions.disabled' to store an array of disabled extension ids. Disabled extensions are loaded but not activated on startup (with a log message).

When extensions are manually deactivated they're added to the list. Since shutdown and reload also deactivate extensions, a system argument is used to differentiate between the two.

When extensions are (manually) started they are removed from the list before trying to load.

Also renamed *activation > *activating so that the log is consistent.

### Screenshot/screencast of this PR

N/A.

### What issues does this PR fix or reference?

Fixes issue #268.

### How to test this PR?

Go to Settings > Extensions and stop an extension. Restart to confirm extension is still stopped. Start it; stop one or two others; restart and check again. No new errors in the log.

jeffmaury reviewed (commented):
I had OpenShift Local extension installed and it says default extension, cannot be removed
Not related to this PR as it is also on main

benoitf commented on packages/main/src/plugin/extension-loader.ts:
```diff
@@ -212,6 +214,22 @@ export class ExtensionLoader {
+        [ExtensionSettings.SectionName + '.' + ExtensionSettings.Disabled]: {
+          description: 'Disabled extensions',
+          type: 'array',
+          default: undefined,
```
should default be [] instead of undefined ?

benoitf commented on packages/main/src/plugin/extension-loader.ts:
```diff
@@ -1036,7 +1075,7 @@ export class ExtensionLoader {
   }
 
-  async deactivateExtension(extensionId: string): Promise<void> {
+  async deactivateExtension(extensionId: string, system?: boolean): Promise<void> {
```
quickly looking at the `system` parameter I don't understand what it will do. Like is it because it's a system extension ?

maybe rename it to `shutdown` or `systemShutdown` or `systemOrder` ?

or just what is triggering the call: `stopAllExtensions`

benoitf commented on packages/main/src/plugin/extension-loader.ts:
```diff
@@ -1047,11 +1086,26 @@ export class ExtensionLoader {
 
+    if (!system) {
+      // extension was explicitly disabled, save the configuration
+      let disabledExtensions = this.getDisabledExtensionIds();
```
could help if the method return always array (including empty array) and not undefined

benoitf commented on packages/main/src/plugin/extension-loader.ts:
```diff
@@ -1047,11 +1086,26 @@ export class ExtensionLoader {
+      // extension was explicitly disabled, save the configuration
+      let disabledExtensions = this.getDisabledExtensionIds();
+
+      if (!disabledExtensions?.includes(extensionId)) {
+        if (disabledExtensions) {
+          disabledExtensions.push(extensionId);
+        } else {
+          disabledExtensions = [extensionId];
+        }
```
maybe we should use a Set instead of Array if you only want at more one element

benoitf reviewed (commented):
to me, a disabled extension is an extension that is not started, not an extension that is not activated.

For example if extension contribute menus, etc. I don't want to see these items. (because it's not only the call to activate function)

I want 0 code of this extension being shown in Podman Desktop

deboer-tim commented:
> I had OpenShift Local extension installed and it says default extension, cannot be removed
> Not related to this PR as it is also on main

I had seen this yesterday, but ran out of time to investigate and confirm it wasn't just my workspace. Confirmed they switch after first restart, opened issue 3149 and will look into it separately.

deboer-tim commented:
> to me, a disabled extension is an extension that is not started, not an extension that is not activated.
> 
> For example if extension contribute menus, etc. I don't want to see these items. (because it's not only the call to activate function)
> 
> I want 0 code of this extension being shown in Podman Desktop

100% agreed. I was not taking a close enough look at what is happening in loadExtension() and it will need to be refactored. Currently it is adding contributions before registering that the extension exists, then trying to activate it. It should register first, check if disabled, then add contributions and start. I will refactor.

benoitf commented:
@deboer-tim I think you can your your filter here https://github.com/containers/podman-desktop/blob/main/packages/main/src/plugin/extension-loader.ts#L343-L354

you remove disabled extensions from the extensions array to be loaded

deboer-tim commented:
I tried that first - doesn't work b/c the extension isn't added to analyzedExtensions until later (https://github.com/containers/podman-desktop/blob/7ad645558d8c9d5bedbc568200068fd569399b6f/packages/main/src/plugin/extension-loader.ts#L533), so it doesn't appear on the extensions page and you can never start it again.

That's the refactoring I think I need to do: if the extension is valid register it as a (stopped) analyzed extension first, before deciding whether to start/adding contributions/etc.

benoitf commented:
ah yes correct, need to add a 'state' field to the AnalyzedExtension 👍

deboer-tim commented:
Updated branch pushed, ready for new review.

Refactored previous commit to have clean extension states:
- Adding an extension to analyzedExtensions is 'registering it with the system'. On startup we load and register all known extensions, then trigger activation for all that haven't been disabled.
- activate/deactivateExtension() are how extensions get started & stopped. Activate registers the contributions (refactored out of loadExtension to new method) and calls it's main; deactivate does the reverse.
- start/stopExtension are how the UI/user requests extensions to be activated or deactivated. stopExtension is new and allows us to differentiate between the system deactivating an extension (e.g. during shutdown or reload) vs the user.

With this refactoring it is much cleaner in that everything to do with loading and unloading an extension is in activate/deactivate (the 'implementation' of starting and stopping), and all external requests come through start/stop (the same as what was already exposed in the API). I've tested all the paths I can think of and no longer see contributions for disabled/stopped extensions.

Additional notes:
- The configuration registry does not support sets. I spent a while trying to add it but haven't been successful yet since you have to write replacer/reviver functions for JSON.stringify/parse. We can either live with an array for now or block this until it supports sets.
- Likewise, if you use `default: []` the configuration registry won't save the property at all. I've narrowed this down and will open a bug, but this PR uses `| []` on line 275 instead.
- Changed the test from a failing activation function to a missing function. This wasn't caught before and would cause no extensions to load, so I extended the try block to catch this.

benoitf reviewed (commented):
it looks like it is changing the behavior of a lifecycle of an extension
Like contributions being part of the 'activate' workflow while the contributions should be added when we load the extension. An extension can be loaded but not activated ( now it's auto-activated) but activation is independent of the load of an extension where we load everything that is contributed

maybe I miss something in a quick look at the code ?

deboer-tim commented:
@benoitf Could you please give another review?

benoitf commented on packages/main/src/plugin/extension-loader.ts:
```diff
@@ -562,9 +567,12 @@ export class ExtensionLoader {
-
-      await this.activateExtension(extension, runtime);
+      // activate the extension if it hasn't been disabled
+      if (!this.getDisabledExtensionIds().includes(extension.id)) {
```
it doesn't feel natural to store disabled state in a separate map/array

we already map of extensions, ids, etc.

Probably the state should go to the extension object directly. It would avoid another indirection level and keep consistency between extensions and the disabled extensions.

I would avoid usage  of `double negation` like instead of not disabled ((!this.getDisabledExtensionIds), check enabled and if state of enablement is in the extension it's easier.

like `if (extension.isEnabled() {
   await this.activateExtension()
`

benoitf commented on packages/main/src/plugin/extension-loader.ts:
```diff
@@ -562,9 +567,12 @@ export class ExtensionLoader {
+        console.log('Not activating disabled extension (' + extension.id + ')');
+      }
     } catch (err) {
       this.extensionState.set(extension.id, 'failed');
       this.extensionStateErrors.set(extension.id, err);
```
state should also go to the extension object

benoitf commented on packages/main/src/plugin/extension-loader.ts:
```diff
@@ -1091,6 +1132,18 @@ export class ExtensionLoader {
 
+  async stopExtension(extensionId: string): Promise<void> {
+    // the user explicitly disabled the extension, save the configuration
+    const disabledExtensions = this.getDisabledExtensionIds();
+
+    if (!disabledExtensions.includes(extensionId)) {
+      disabledExtensions.push(extensionId);
+      this.setDisabledExtensionIds(disabledExtensions);
+    }
+
```
here it would be easier if disabled is already stored in the extension.

it's strange also that we don't check if extension exists before adding it to the list ?

benoitf replied:
I'm not sure to understand also what it does.

if I stop and extension and the extension is not in the disabled list, it adds this extension as being disabled

so it is not a stop but a disableExtension method

benoitf commented on packages/main/src/plugin/extension-loader.ts:
```diff
@@ -1100,6 +1153,16 @@ export class ExtensionLoader {
   async startExtension(extensionId: string): Promise<void> {
     const extension = this.analyzedExtensions.get(extensionId);
     if (extension) {
+      const disabledExtensionIds = this.getDisabledExtensionIds();
+      const index = disabledExtensionIds.indexOf(extensionId);
+      if (index > -1) {
+        // extension was disabled but the user has activated it,
+        // clean the configuration up
+        disabledExtensionIds.splice(index, 1);
+
+        this.setDisabledExtensionIds(disabledExtensionIds);
+      }
+
```
if state is in the extension we do only
`if (extension.isEnabled()...)`

benoitf commented on packages/main/src/plugin/index.ts:
```diff
@@ -1349,9 +1349,9 @@ export class PluginSystem {
     this.ipcHandle(
-      'extension-loader:deactivateExtension',
+      'extension-loader:stopExtension',
       async (_listener: Electron.IpcMainInvokeEvent, extensionId: string): Promise<void> => {
-        return this.extensionLoader.deactivateExtension(extensionId);
+        return this.extensionLoader.stopExtension(extensionId);
```
I guess it's disableExtension now ?

benoitf commented on packages/preload/src/index.ts:
```diff
@@ -942,7 +942,7 @@ function initExposure(): void {
   });
 
   contextBridge.exposeInMainWorld('stopExtension', async (extensionId: string): Promise<void> => {
-    return ipcInvoke('extension-loader:deactivateExtension', extensionId);
+    return ipcInvoke('extension-loader:stopExtension', extensionId);
```
disableExtension

benoitf reviewed (commented):
I think state was added later (as a separate map), then stateError was added (also as a separate map) and now you're trying to add enable/disable as a separate array.

I think as you're switching to always having extensions (to get state/errors/etc), everything needs to be merged in a single object.

for example when we list extensions we do
```
async listExtensions(): Promise<ExtensionInfo[]> {
    return Array.from(this.analyzedExtensions.values()).map(extension => ({
      state: this.extensionState.get(extension.id) || 'stopped',
      error: this.mapError(this.extensionStateErrors.get(extension.id)),

```

it should not use for state and error another data, it should already be there, in the maintained Map


so in short, one map with extensionId as a key

then state, enabled/disabled, errors, manifest, etc it's all in the same object.

Else it comes to be a nightmare as we need to maintain state across 4 different maps/arrays in the same Class for the same stuff

cdrage commented:
Just going through some old PR's, @deboer-tim what's the status of the PR?
